diff --git a/build/AgentProvider.d.ts b/build/AgentProvider.d.ts
index 1c1c0c867f21d899f62bf99725925ffe0084be91..8e654fc25fdc71d098088088235ee4782e894271 100644
--- a/build/AgentProvider.d.ts
+++ b/build/AgentProvider.d.ts
@@ -3,11 +3,15 @@ import type { PropsWithChildren } from 'react';
 import * as React from 'react';
 interface AgentContextInterface<AppAgent extends Agent = Agent> {
     loading: boolean;
-    agent: AppAgent;
+    agent?: AppAgent;
+    publicDid?: string;
+    setAgent(agent: AppAgent | null): void;
+    setPublicDid(did: string): void;
+    agentIteration: number;
 }
 export declare const useAgent: <AppAgent extends Agent<any>>() => AgentContextInterface<AppAgent>;
 interface Props {
-    agent: Agent;
+    agent: Agent | undefined;
 }
 declare const AgentProvider: React.FC<PropsWithChildren<Props>>;
 export default AgentProvider;
\ No newline at end of file
diff --git a/build/AgentProvider.js b/build/AgentProvider.js
index 2f4560a13ca0271da8c8178a269a3288768eed5a..d30850cc488020ed12a52ebc9c6b06b3b643e5a8 100644
--- a/build/AgentProvider.js
+++ b/build/AgentProvider.js
@@ -48,18 +48,26 @@ const useAgent = () => {
 };
 exports.useAgent = useAgent;
 const AgentProvider = ({ agent, children }) => {
-    const isQaRegistered = (0, recordUtils_1.useIsModuleRegistered)(agent, question_answer_1.QuestionAnswerModule);
-    const [agentState] = (0, react_1.useState)({
-        loading: false,
+    const [agentState, setAgentState] = (0, react_1.useState)({
+        loading: true,
         agent,
+        agentIteration: 1,
+        publicDid: undefined
     });
-    return (React.createElement(AgentContext.Provider, { value: agentState },
-        React.createElement(ConnectionProvider_1.default, { agent: agent },
-            React.createElement(CredentialProvider_1.default, { agent: agent },
-                React.createElement(ProofProvider_1.default, { agent: agent },
-                    React.createElement(CredentialFormatDataProvider_1.default, { agent: agent },
-                        React.createElement(ProofFormatDataProvider_1.default, { agent: agent },
-                            React.createElement(BasicMessageProvider_1.default, { agent: agent }, isQaRegistered ? (React.createElement(QuestionAnswerProvider_1.default, { agent: agent },
+    const isQaRegistered = (0, recordUtils_1.useIsModuleRegistered)(agentState.agent, question_answer_1.QuestionAnswerModule);
+    const setInitialState = async (newAgent) => {
+        setAgentState(prev => ({ agent: newAgent, loading: !newAgent, agentIteration: newAgent ? prev.agentIteration + 1 : prev.agentIteration, publicDid: prev.publicDid }));
+    };
+    const setPublicDid = (publicDid) => {
+        setAgentState((state) => ({ ...state, publicDid }))
+    }
+    return (React.createElement(AgentContext.Provider, { value: { agent: agentState.agent, publicDid: agentState.publicDid, agentIteration: agentState.agentIteration, loading: agentState.agent ? false : true, setAgent: setInitialState, setPublicDid } },
+        React.createElement(ConnectionProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration },
+            React.createElement(CredentialProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration },
+                React.createElement(ProofProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration },
+                    React.createElement(CredentialFormatDataProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration },
+                        React.createElement(ProofFormatDataProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration },
+                            React.createElement(BasicMessageProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration }, isQaRegistered ? (React.createElement(QuestionAnswerProvider_1.default, { agent: agentState.agent, agentIteration: agentState.agentIteration },
                                 children,
                                 " ")) : (children)))))))));
 };
diff --git a/build/BasicMessageProvider.d.ts b/build/BasicMessageProvider.d.ts
index c94912b304da423451509d11d90eb443b632e40e..c871e56e0d47c9c4dafc679268849aa3a78f8356 100644
--- a/build/BasicMessageProvider.d.ts
+++ b/build/BasicMessageProvider.d.ts
@@ -6,7 +6,8 @@ import { BifoldAgent } from './agent';
 export declare const useBasicMessages: () => RecordsState<DidCommBasicMessageRecord>;
 export declare const useBasicMessagesByConnectionId: (connectionId: string) => DidCommBasicMessageRecord[];
 interface Props {
-    agent: BifoldAgent;
+    agent: BifoldAgent | undefined;
+    agentIteration: number;
 }
 declare const BasicMessageProvider: React.FC<PropsWithChildren<Props>>;
 export default BasicMessageProvider;
diff --git a/build/BasicMessageProvider.js b/build/BasicMessageProvider.js
index b718e212bd00f887d314b977e09a1ac4ccb13b58..79644401b5c3c6919a7d7d02f2e292777e0f190f 100644
--- a/build/BasicMessageProvider.js
+++ b/build/BasicMessageProvider.js
@@ -43,12 +43,15 @@ const useBasicMessagesByConnectionId = (connectionId) => {
     return messages;
 };
 exports.useBasicMessagesByConnectionId = useBasicMessagesByConnectionId;
-const BasicMessageProvider = ({ agent, children }) => {
+const BasicMessageProvider = ({ agent, agentIteration, children }) => {
     const [state, setState] = (0, react_1.useState)({
         records: [],
         loading: true,
     });
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const records = await agent.modules.didcomm.basicMessages.findAllByQuery({});
         setState({ records, loading: false });
     }, [agent]);
@@ -56,7 +59,7 @@ const BasicMessageProvider = ({ agent, children }) => {
         setInitialState();
     }, [setInitialState]);
     (0, react_1.useEffect)(() => {
-        if (state.loading)
+        if (state.loading || !agent)
             return;
         const basicMessageAdded$ = (0, recordUtils_1.recordsAddedByType)(agent, didcomm_1.DidCommBasicMessageRecord).subscribe((record) => setState((0, recordUtils_1.addRecord)(record, state)));
         const basicMessageUpdated$ = (0, recordUtils_1.recordsUpdatedByType)(agent, didcomm_1.DidCommBasicMessageRecord).subscribe((record) => setState((0, recordUtils_1.updateRecord)(record, state)));
@@ -66,7 +69,7 @@ const BasicMessageProvider = ({ agent, children }) => {
             basicMessageUpdated$ === null || basicMessageUpdated$ === void 0 ? void 0 : basicMessageUpdated$.unsubscribe();
             basicMessageRemoved$ === null || basicMessageRemoved$ === void 0 ? void 0 : basicMessageRemoved$.unsubscribe();
         };
-    }, [state, agent]);
+    }, [state, agent, agentIteration]);
     return React.createElement(BasicMessageContext.Provider, { value: state }, children);
 };
 exports.default = BasicMessageProvider;
diff --git a/build/ConnectionProvider.d.ts b/build/ConnectionProvider.d.ts
index 40ffa7eb50c538dac696cf9fcc253ac4785360dd..d0921a9946734934e4062d0cb682136d1533ff24 100644
--- a/build/ConnectionProvider.d.ts
+++ b/build/ConnectionProvider.d.ts
@@ -18,7 +18,8 @@ export declare const useConnections: (options?: UseConnectionsOptions) => {
 };
 export declare const useConnectionById: (id: string) => DidCommConnectionRecord | undefined;
 interface Props {
-    agent: BifoldAgent;
+    agent: BifoldAgent | undefined;
+    agentIteration: number;
 }
 declare const ConnectionProvider: React.FC<PropsWithChildren<Props>>;
 export default ConnectionProvider;
diff --git a/build/ConnectionProvider.js b/build/ConnectionProvider.js
index 3f89fdbdfcae0275465496cb419c819493fc87fb..bca8a8cfec4264ec0fc7c2ccab04b2faa6747509 100644
--- a/build/ConnectionProvider.js
+++ b/build/ConnectionProvider.js
@@ -66,12 +66,15 @@ const useConnectionById = (id) => {
     return connections.find((c) => c.id === id);
 };
 exports.useConnectionById = useConnectionById;
-const ConnectionProvider = ({ agent, children }) => {
+const ConnectionProvider = ({ agent, agentIteration, children }) => {
     const [state, setState] = (0, react_1.useState)({
         records: [],
         loading: true,
     });
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const records = await agent.modules.didcomm.connections.getAll();
         setState({ records, loading: false });
     }, [agent]);
@@ -79,7 +82,7 @@ const ConnectionProvider = ({ agent, children }) => {
         setInitialState();
     }, [setInitialState]);
     (0, react_1.useEffect)(() => {
-        if (state.loading)
+        if (state.loading || !agent)
             return;
         const connectionAdded$ = (0, recordUtils_1.recordsAddedByType)(agent, didcomm_1.DidCommConnectionRecord).subscribe((record) => setState((0, recordUtils_1.addRecord)(record, state)));
         const connectionUpdated$ = (0, recordUtils_1.recordsUpdatedByType)(agent, didcomm_1.DidCommConnectionRecord).subscribe((record) => setState((0, recordUtils_1.updateRecord)(record, state)));
@@ -89,7 +92,7 @@ const ConnectionProvider = ({ agent, children }) => {
             connectionUpdated$.unsubscribe();
             connectionRemoved$.unsubscribe();
         };
-    }, [state, agent]);
+    }, [state, agent, agentIteration]);
     return React.createElement(ConnectionContext.Provider, { value: state }, children);
 };
 exports.default = ConnectionProvider;
diff --git a/build/CredentialFormatDataProvider.d.ts b/build/CredentialFormatDataProvider.d.ts
index 732edf69828eb9efb5c06e6d717775ed6a85f128..aff96fdab9dfa0d294edb3b898819053bf11d76e 100644
--- a/build/CredentialFormatDataProvider.d.ts
+++ b/build/CredentialFormatDataProvider.d.ts
@@ -12,7 +12,8 @@ type FormattedDataState = {
 export declare const useCredentialsFormatData: () => FormattedDataState;
 export declare const useCredentialFormatDataById: (id: string) => CredentialFormatData | undefined;
 interface Props {
-    agent: BifoldAgent;
+    agent: BifoldAgent | undefined;
+    agentIteration: number;
 }
 declare const CredentialFormatDataProvider: React.FC<PropsWithChildren<Props>>;
 export default CredentialFormatDataProvider;
diff --git a/build/CredentialFormatDataProvider.js b/build/CredentialFormatDataProvider.js
index 3102ed6279a61cda9ed53ed0dbe38e08f5e42232..b417bd3f912682088d60d61066ac46743699703f 100644
--- a/build/CredentialFormatDataProvider.js
+++ b/build/CredentialFormatDataProvider.js
@@ -67,7 +67,7 @@ const useCredentialFormatDataById = (id) => {
     return formattedData.find((c) => c.id === id);
 };
 exports.useCredentialFormatDataById = useCredentialFormatDataById;
-const CredentialFormatDataProvider = ({ agent, children }) => {
+const CredentialFormatDataProvider = ({ agent, agentIteration, children }) => {
     const [state, setState] = (0, react_1.useState)({
         formattedData: [],
         loading: true,
@@ -77,6 +77,9 @@ const CredentialFormatDataProvider = ({ agent, children }) => {
         return Object.assign(Object.assign({}, formatData), { id: record.id });
     };
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const records = await agent.modules.didcomm.credentials.getAll();
         const formattedData = [];
         for (const record of records) {
@@ -88,13 +91,17 @@ const CredentialFormatDataProvider = ({ agent, children }) => {
         void setInitialState();
     }, [setInitialState]);
     (0, react_1.useEffect)(() => {
-        if (state.loading)
+        if (state.loading || !agent)
             return;
         const credentialAdded$ = (0, recordUtils_1.recordsAddedByType)(agent, didcomm_1.DidCommCredentialExchangeRecord).subscribe(async (record) => {
+            if (!agent)
+                return;
             const formatData = await fetchCredentialInformation(agent, record);
             setState(addRecord(formatData, state));
         });
         const credentialUpdate$ = (0, recordUtils_1.recordsUpdatedByType)(agent, didcomm_1.DidCommCredentialExchangeRecord).subscribe(async (record) => {
+            if (!agent)
+                return;
             const formatData = await fetchCredentialInformation(agent, record);
             setState(updateRecord(formatData, state));
         });
@@ -104,7 +111,7 @@ const CredentialFormatDataProvider = ({ agent, children }) => {
             credentialUpdate$.unsubscribe();
             credentialRemove$.unsubscribe();
         };
-    }, [state, agent]);
+    }, [state, agent, agentIteration]);
     return react_1.default.createElement(CredentialFormatDataContext.Provider, { value: state }, children);
 };
 exports.default = CredentialFormatDataProvider;
diff --git a/build/CredentialProvider.d.ts b/build/CredentialProvider.d.ts
index 60e12ff5b610c183ec7c3341fb8ec593ad9ba8cd..da3fde6b721f29fb56e14f0b9a298c26d7698b9e 100644
--- a/build/CredentialProvider.d.ts
+++ b/build/CredentialProvider.d.ts
@@ -9,7 +9,8 @@ export declare const useCredentialById: (id: string) => DidCommCredentialExchang
 export declare const useCredentialByState: (state: DidCommCredentialState | DidCommCredentialState[]) => DidCommCredentialExchangeRecord[];
 export declare const useCredentialNotInState: (state: DidCommCredentialState | DidCommCredentialState[]) => DidCommCredentialExchangeRecord[];
 interface Props {
-    agent: Agent;
+    agent: Agent | undefined;
+    agentIteration: number;
 }
 declare const CredentialProvider: React.FC<PropsWithChildren<Props>>;
 export default CredentialProvider;
diff --git a/build/CredentialProvider.js b/build/CredentialProvider.js
index 5db95be63a60633b0fdb5502a15353e51e8ddbce..ba3e401f380d1361a94250b59015f64b42be4779 100644
--- a/build/CredentialProvider.js
+++ b/build/CredentialProvider.js
@@ -61,12 +61,15 @@ const useCredentialNotInState = (state) => {
     return filteredCredentials;
 };
 exports.useCredentialNotInState = useCredentialNotInState;
-const CredentialProvider = ({ agent, children }) => {
+const CredentialProvider = ({ agent, agentIteration, children }) => {
     const [state, setState] = (0, react_1.useState)({
         records: [],
         loading: true,
     });
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const records = await agent.modules.didcomm.credentials.getAll();
         setState({ records, loading: false });
     }, [agent]);
@@ -74,7 +77,7 @@ const CredentialProvider = ({ agent, children }) => {
         setInitialState();
     }, [setInitialState]);
     (0, react_1.useEffect)(() => {
-        if (state.loading)
+        if (state.loading || !agent)
             return;
         const credentialAdded$ = (0, recordUtils_1.recordsAddedByType)(agent, didcomm_1.DidCommCredentialExchangeRecord).subscribe((record) => setState((0, recordUtils_1.addRecord)(record, state)));
         const credentialUpdated$ = (0, recordUtils_1.recordsUpdatedByType)(agent, didcomm_1.DidCommCredentialExchangeRecord).subscribe((record) => setState((0, recordUtils_1.updateRecord)(record, state)));
@@ -84,7 +87,7 @@ const CredentialProvider = ({ agent, children }) => {
             credentialUpdated$ === null || credentialUpdated$ === void 0 ? void 0 : credentialUpdated$.unsubscribe();
             credentialRemoved$ === null || credentialRemoved$ === void 0 ? void 0 : credentialRemoved$.unsubscribe();
         };
-    }, [state, agent]);
+    }, [state, agent, agentIteration]);
     return React.createElement(CredentialContext.Provider, { value: state }, children);
 };
 exports.default = CredentialProvider;
diff --git a/build/ProofFormatDataProvider.d.ts b/build/ProofFormatDataProvider.d.ts
index 382b50e912f37804fb6e8387841500b044057085..d8bcce2fb0473822906403eeb16e0801517b7bfa 100644
--- a/build/ProofFormatDataProvider.d.ts
+++ b/build/ProofFormatDataProvider.d.ts
@@ -13,7 +13,8 @@ type FormattedProofDataState = {
 export declare const useProofsFormatData: () => FormattedProofDataState;
 export declare const useProofFormatDataById: (id: string) => ProofFormatData | undefined;
 interface Props {
-    agent: Agent;
+    agent: Agent | undefined;
+    agentIteration: number;
 }
 declare const ProofFormatDataProvider: React.FC<PropsWithChildren<Props>>;
 export default ProofFormatDataProvider;
diff --git a/build/ProofFormatDataProvider.js b/build/ProofFormatDataProvider.js
index d7e52d4c099b209a3c86832136efde60a9932e00..07a20df135a815775cca6ba2bacdca333d1f36f2 100644
--- a/build/ProofFormatDataProvider.js
+++ b/build/ProofFormatDataProvider.js
@@ -68,12 +68,15 @@ const useProofFormatDataById = (id) => {
     return formattedData.find((c) => c.id === id);
 };
 exports.useProofFormatDataById = useProofFormatDataById;
-const ProofFormatDataProvider = ({ agent, children }) => {
+const ProofFormatDataProvider = ({ agent, agentIteration, children }) => {
     const [state, setState] = (0, react_1.useState)({
         formattedData: [],
         loading: true,
     });
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const records = await agent.modules.didcomm.proofs.getAll();
         const formattedData = [];
         for (const record of records) {
@@ -86,13 +89,17 @@ const ProofFormatDataProvider = ({ agent, children }) => {
         void setInitialState();
     }, [setInitialState]);
     (0, react_1.useEffect)(() => {
-        if (state.loading)
+        if (state.loading || !agent)
             return;
         const proofAdded$ = (0, recordUtils_1.recordsAddedByType)(agent, didcomm_1.DidCommProofExchangeRecord).subscribe(async (record) => {
+            if (!agent)
+                return;
             const formatData = await agent.modules.didcomm.proofs.getFormatData(record.id);
             setState(addRecord(Object.assign(Object.assign({}, formatData), { id: record.id }), state));
         });
         const proofUpdate$ = (0, recordUtils_1.recordsUpdatedByType)(agent, didcomm_1.DidCommProofExchangeRecord).subscribe(async (record) => {
+            if (!agent)
+                return;
             const formatData = await agent.modules.didcomm.proofs.getFormatData(record.id);
             setState(updateRecord(Object.assign(Object.assign({}, formatData), { id: record.id }), state));
         });
@@ -102,7 +109,7 @@ const ProofFormatDataProvider = ({ agent, children }) => {
             proofUpdate$.unsubscribe();
             proofRemove$.unsubscribe();
         };
-    }, [state, agent]);
+    }, [state, agent, agentIteration]);
     return React.createElement(ProofFormatDataContext.Provider, { value: state }, children);
 };
 exports.default = ProofFormatDataProvider;
diff --git a/build/ProofProvider.d.ts b/build/ProofProvider.d.ts
index d4fc0bc5a61aab48970180f41ca356e8ff8f5deb..d992a28290d451403c068db199905b30bb7b4791 100644
--- a/build/ProofProvider.d.ts
+++ b/build/ProofProvider.d.ts
@@ -9,7 +9,8 @@ export declare const useProofById: (id: string) => DidCommProofExchangeRecord |
 export declare const useProofByState: (state: DidCommProofState | DidCommProofState[]) => DidCommProofExchangeRecord[];
 export declare const useProofNotInState: (state: DidCommProofState | DidCommProofState[]) => DidCommProofExchangeRecord[];
 interface Props {
-    agent: BifoldAgent;
+    agent: BifoldAgent | undefined;
+    agentIteration: number;
 }
 declare const ProofProvider: React.FC<PropsWithChildren<Props>>;
 export default ProofProvider;
diff --git a/build/ProofProvider.js b/build/ProofProvider.js
index 82dd40ad2ae240c10abf0ec447ab2668c0da3854..a0153ac444f9e1f0a89f85c92a179518750b424b 100644
--- a/build/ProofProvider.js
+++ b/build/ProofProvider.js
@@ -67,12 +67,15 @@ const useProofNotInState = (state) => {
     return filteredProofs;
 };
 exports.useProofNotInState = useProofNotInState;
-const ProofProvider = ({ agent, children }) => {
+const ProofProvider = ({ agent, agentIteration, children }) => {
     const [state, setState] = (0, react_1.useState)({
         records: [],
         loading: true,
     });
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const records = await agent.modules.didcomm.proofs.getAll();
         setState({ records, loading: false });
     }, [agent]);
@@ -80,7 +83,7 @@ const ProofProvider = ({ agent, children }) => {
         setInitialState();
     }, [setInitialState]);
     (0, react_1.useEffect)(() => {
-        if (state.loading)
+        if (state.loading || !agent)
             return;
         const proofAdded$ = (0, recordUtils_1.recordsAddedByType)(agent, didcomm_1.DidCommProofExchangeRecord).subscribe((record) => setState((0, recordUtils_1.addRecord)(record, state)));
         const proofUpdated$ = (0, recordUtils_1.recordsUpdatedByType)(agent, didcomm_1.DidCommProofExchangeRecord).subscribe((record) => setState((0, recordUtils_1.updateRecord)(record, state)));
@@ -90,7 +93,7 @@ const ProofProvider = ({ agent, children }) => {
             proofUpdated$ === null || proofUpdated$ === void 0 ? void 0 : proofUpdated$.unsubscribe();
             proofRemoved$ === null || proofRemoved$ === void 0 ? void 0 : proofRemoved$.unsubscribe();
         };
-    }, [state, agent]);
+    }, [state, agent, agentIteration]);
     return React.createElement(ProofContext.Provider, { value: state }, children);
 };
 exports.default = ProofProvider;
diff --git a/build/QuestionAnswerProvider.d.ts b/build/QuestionAnswerProvider.d.ts
index 95a6c38861bdf71d46ba145c50000150c664a2f6..f6d918083feaf75a81d2d9bce7f4666ebf4bf710 100644
--- a/build/QuestionAnswerProvider.d.ts
+++ b/build/QuestionAnswerProvider.d.ts
@@ -8,7 +8,8 @@ export declare const useQuestionAnswer: () => {
 export declare const useQuestionAnswerByConnectionId: (connectionId: string) => QuestionAnswerRecord[];
 export declare const useQuestionAnswerById: (id: string) => QuestionAnswerRecord | undefined;
 interface Props {
-    agent: Agent;
+    agent: Agent | undefined;
+    agentIteration: number;
 }
 declare const QuestionAnswerProvider: React.FC<PropsWithChildren<Props>>;
 export default QuestionAnswerProvider;
diff --git a/build/QuestionAnswerProvider.js b/build/QuestionAnswerProvider.js
index 1b86523c16e5e7b10a01af57378e7f4e55f573e0..eb24523d223943215d62196c192a362c8fc90c62 100644
--- a/build/QuestionAnswerProvider.js
+++ b/build/QuestionAnswerProvider.js
@@ -47,12 +47,15 @@ const useQuestionAnswerById = (id) => {
     return questionAnswerMessages.find((c) => c.id === id);
 };
 exports.useQuestionAnswerById = useQuestionAnswerById;
-const QuestionAnswerProvider = ({ agent, children }) => {
+const QuestionAnswerProvider = ({ agent, agentIteration, children }) => {
     const [questionAnswerState, setQuestionAnswerState] = (0, react_1.useState)({
         questionAnswerMessages: [],
         loading: true,
     });
     const setInitialState = (0, react_1.useCallback)(async () => {
+        setQuestionAnswerState(prev => ({ ...prev, loading: true }));
+        if (!agent)
+            return;
         const questAnswerApi = agent.dependencyManager.resolve(question_answer_1.QuestionAnswerApi);
         const questionAnswerMessages = await questAnswerApi.getAll();
         setQuestionAnswerState({ questionAnswerMessages, loading: false });
@@ -77,11 +80,11 @@ const QuestionAnswerProvider = ({ agent, children }) => {
                 questionAnswerMessages: newQuestionAnswerState,
             });
         };
-        agent.events.on(question_answer_1.QuestionAnswerEventTypes.QuestionAnswerStateChanged, listener);
+        agent === null || agent === void 0 ? void 0 : agent.events.on(question_answer_1.QuestionAnswerEventTypes.QuestionAnswerStateChanged, listener);
         return () => {
-            agent.events.off(question_answer_1.QuestionAnswerEventTypes.QuestionAnswerStateChanged, listener);
+            agent === null || agent === void 0 ? void 0 : agent.events.off(question_answer_1.QuestionAnswerEventTypes.QuestionAnswerStateChanged, listener);
         };
-    }, [questionAnswerState, agent]);
+    }, [questionAnswerState, agent, agentIteration]);
     return React.createElement(QuestionAnswerContext.Provider, { value: questionAnswerState }, children);
 };
 exports.default = QuestionAnswerProvider;
diff --git a/build/recordUtils.d.ts b/build/recordUtils.d.ts
index 40de6e6278668145b77ea69d86d7b006e14e0b4e..bfc59eab2c4d1eaaf994da5585b27c1a43f7783c 100644
--- a/build/recordUtils.d.ts
+++ b/build/recordUtils.d.ts
@@ -17,5 +17,5 @@ export declare const recordsAddedByType: <R extends BaseRecordAny>(agent: Agent
 export declare const recordsUpdatedByType: <R extends BaseRecordAny>(agent: Agent | undefined, recordClass: RecordClass<R>) => import("rxjs").Observable<R>;
 export declare const recordsRemovedByType: <R extends BaseRecordAny>(agent: Agent | undefined, recordClass: RecordClass<R>) => import("rxjs").Observable<R>;
 export declare const isModuleRegistered: (agent: Agent, ModuleClass: Constructor) => boolean;
-export declare const useIsModuleRegistered: (agent: Agent, ModuleClass: Constructor) => boolean;
+export declare const useIsModuleRegistered: (agent: Agent | undefined, ModuleClass: Constructor) => boolean | undefined;
 export {};
diff --git a/build/recordUtils.js b/build/recordUtils.js
index 5fcf63af4d1790d4de9c93a7c5726e0d97fdf491..e6a5a553877230a09598cd6eb32e3274ae6f1699 100644
--- a/build/recordUtils.js
+++ b/build/recordUtils.js
@@ -75,7 +75,7 @@ const isModuleRegistered = (agent, ModuleClass) => {
 };
 exports.isModuleRegistered = isModuleRegistered;
 const useIsModuleRegistered = (agent, ModuleClass) => {
-    return (0, react_1.useMemo)(() => (0, exports.isModuleRegistered)(agent, ModuleClass), [agent, ModuleClass]);
+    return (0, react_1.useMemo)(() => (agent ? (0, exports.isModuleRegistered)(agent, ModuleClass) : undefined), [agent, ModuleClass]);
 };
 exports.useIsModuleRegistered = useIsModuleRegistered;
 //# sourceMappingURL=recordUtils.js.map
\ No newline at end of file
